### PR TITLE
feat(spans): Add a metrics summaries dataset

### PR DESCRIPTION
**Pull request by phacops**
For DDM, we'd like to add a way to provide examples of spans which generated some metrics. The idea is, when we're visualizing a DDM metric, the user would be able to select a square on the chart and we'd show them a list of spans associated with those values. This would add a new dataset called `metrics_summaries` in order to correlate DDM metrics with spans.

In order to correlate DDM metrics and spans, [RFC 123](https://github.com/getsentry/rfcs/blob/main/text/0123-metrics-correlation.md) adds a metrics summary to spans. The relationship between spans and summaries is like this:
```
1 span -> M metrics -> S metrics summaries -> T tags
```

To display metrics information when querying spans, we'd like to add a column called `metrics_summary` to the `spans` dataset to purely **store** the needed information. When rendering those spans, we'd be able to show information about the metrics and the different values.

In order to do the reverse, showing spans related to some specific metrics, we'd need a way to filter through these spans' metrics summaries. Given the multi-dimensional nature of metrics summaries (we could have many summaries attached to 1 metric attached to 1 span), using the same way we do to filter on `tags` is not possible without flattening the tags as well and it's less efficient to fully filter that way.
One efficient way to do this is have a separate dataset. We'd store enough information about the spans so we don't have to join (in ClickHouse or in the app). We could keep tags as an array column or even insert 1 row per tag value. This way, selecting a list of spans corresponding to certain values for given metric would be a simple `SELECT` query away.

### CODE REVIEW

**Comment by evanh** [comment]
Some high level questions:

- Have you done any analysis on the size/type of data you'll be storing in this table?
- How are you measuring the success of this? How do you know what part of this solution is working/not working?
- Is there any documentation of the decisions that lead to this particular solution?
- Is there some documentation on what the product use cases of this are?
    - On a related note, I'm not sure I understand the utility of seeing the metrics associated with a specific span. Metrics are generally useful in aggregate, and the metrics attached to one singular span would basically be raw values.
    - I think the more complex problem is actually the second one you mentioned, going from metrics -> spans. That will require a completely different table layout than what you propose here.

File: snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+from snuba.migrations.operations import OperationTarget, SqlOperation
+from snuba.utils.schemas import Float
+
+storage_set_name = StorageSetKey.METRICS_SUMMARIES
+local_table_name = "metrics_summaries_local"
+dist_table_name = "metrics_summaries_dist"
+
+UNKNOWN_SPAN_STATUS = 2
+
+columns: List[Column[Modifiers]] = [
```
Comment (evanh) [inline snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py:16]: This is missing any connection to our actual metrics dataset (metric_id, use case, mri etc.) Is that intended?

File: snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    blocking = False
+
+    def forwards_ops(self) -> Sequence[SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=storage_set_name,
+                table_name=local_table_name,
+                columns=columns,
+                engine=table_engines.ReplacingMergeTree(
+                    order_by="(project_id, group, cityHash64(span_id))",
```
Comment (evanh) [inline snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py:52]: Is every query to this table going to have a `group` and `span_id`?

File: snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    Column("max", Float(64)),
+    Column("sum", Float(64)),
+    Column("count", Float(64)),
+    # metrics metadata
+    Column("tags", Nested([("key", String()), ("value", String())])),
+    # span metadata
+    Column("start_timestamp", DateTime()),
+    Column("exclusive_time", Float(64)),
+    Column("op", String(Modifiers(low_cardinality=True))),
+    Column("group", UInt(64)),
```
Comment (evanh) [inline snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py:33]: What is `group` in this context?

File: snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py (hunk 1 above)
Comment (phacops) [reply snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py:16]: No, that was a mistake.

File: snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py (hunk 3 above)
Comment (phacops) [reply snuba/snuba_migrations/metrics_summaries/0001_metrics_summaries_create_table.py:33]: `group` is one way to select span examples. We might want to target exact spans producing those specific metric values or spans looking like them.

We might be able to not have that and just focus on returning span IDs.

**Comment by phacops** [comment]
As asked by @evanh , I split this PR:
- https://github.com/getsentry/snuba/pull/5159
- https://github.com/getsentry/snuba/pull/5160
- https://github.com/getsentry/snuba/pull/5161

**Pull request closed without merging**